cacheops/redis_client.py

```diff
@@ -0,0 +1,215 @@
+import json
+import os.path
+import re
+from threading import local
+
+from funcy import decorator, identity, memoize
+
+try:
+    from collections import ChainMap
+except ImportError:
+    from chainmap import ChainMap
+
+from redis import ConnectionError, TimeoutError
+from redis.client import StrictRedis
+
+from .cross import pickle
+from .conf import LRU, DEGRADE_ON_FAILURE, REDIS_CONF
+import warnings
+
+_marker = object()
+
+STRIP_RE = re.compile(r'TOSTRIP.*/TOSTRIP', re.S)
+
+# Support DEGRADE_ON_FAILURE
+if DEGRADE_ON_FAILURE:
+    @decorator
+    def handle_connection_failure(call):
+        try:
+            return call()
+        except ConnectionError as e:
+            warnings.warn("The cacheops cache is unreachable! Error: %s" % e, RuntimeWarning)
+        except TimeoutError as e:
+            warnings.warn("The cacheops cache timed out! Error: %s" % e, RuntimeWarning)
+else:
+    handle_connection_failure = identity
+
+
+class CacheMiss(Exception):
+    pass
+
+
+class LocalCachedTransactionRedis(StrictRedis):
+    def __init__(self, *args, **kwargs):
+        super(LocalCachedTransactionRedis, self).__init__(*args, **kwargs)
+        self._local = local()
+
+    @memoize
+    def load_script(self, name, strip=False):
+        """ LUA Script Loader """
+        # TODO: strip comments
+        filename = os.path.join(os.path.dirname(__file__), 'lua/%s.lua' % name)
+        with open(filename) as f:
+            code = f.read()
+        if strip:
+            code = STRIP_RE.sub('', code)
+        return self.register_script(code)
+
+    def get(self, name):
+        # try transaction local cache first
+        try:
+            cache_data = self._local.cacheops_transaction_cache.get(name, None)
+        except AttributeError:
+            # not in transaction
+            pass
+        else:
+            if cache_data is not None and cache_data.get('data', _marker) is not _marker:
+                return cache_data['data']
+        cache_data = super(LocalCachedTransactionRedis, self).get(name)
+        if cache_data is None:
+            raise CacheMiss
+        return pickle.loads(cache_data)
+
+    @handle_connection_failure
+    def cache_thing(self, cache_key, data, cond_dnfs, timeout):
+        """
+        Writes data to cache and creates appropriate invalidators.
+        """
+        try:
+            # are we in a transaction?
+            self._local.cacheops_transaction_cache[cache_key] = {
+                'data': data,
+                'cond_dnfs': cond_dnfs,
+                'timeout': timeout,
+                # these two help us out later for possible invalidation
+                'db_tables': [x for x, y in cond_dnfs],
+                'cond_dicts': [dict(i) for x, y in cond_dnfs for i in y]
+            }
```

Comment:
As local cache is not really needed to support transactions, I suggest removing it for now. It could be added later with separately. This way we will hopefully do less bugs.

Also, there should be better code to check if we are in transaction, like:
```python
if self.in_transaction:
    ...
else:
    ...
```
Should be trivial to add property like that.